crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1113 @@
+    pub fn new(signers: Vec<SignerContainer<S>>, threshold: u32) -> Self {
+        Self { signers, threshold }
+    }
+
+    /// Get the list of potential signers.
+    pub fn signers(&self) -> &[SignerContainer<S>] {
+        &self.signers
+    }
+
+    /// Get the threshold.
+    pub fn threshold(&self) -> u32 {
+        self.threshold
+    }
+
+    /// Verify a message against a multi-signature, returning the list of
+    /// signers that signed it.
+    ///
+    /// Note that a signer is allowed to appear in multiple signer sets. For
+    /// example, assume a signer set that requires 2 out of 2 signers, each
+    /// being its own signer set: 1) SignerSet1: SignerA, SignerB (1 out of
+    /// 2) 2) SignerSet2: SignerA, SignerC (1 out of 2)
+    /// 3) SignerSetX: SignerSet1, SignerSet (2 out of 2)
+    /// When validating SignerSetX, providing just a signature from SignerA will
+    /// be enough to satisfy all thresholds. This is acceptable because it
+    /// means the rules for verifying a single signer set and a nested one
+    /// are the same, and easier to follow. The example above simply shows a
+    /// poorly specified signer set.
+    pub fn verify<SIG>(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<SIG>,
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        SIG: Clone + Default + Digestible + Eq + Hash + Message + Ord + Serialize + Signature,
+        S: Verifier<SIG>,
+    {
+        if multi_sig.signatures().len() > MAX_SIGNATURES {
+            return Err(SignatureError::new());
+        }
+
+        // Sort and dedup the list of signatures. Even though we match signers to
+        // signatures (and not the other way around), we still deuplicate the
+        // list of signatures since it is cheap to do, and a reasonable defensive
+        // programming measure.
+        let mut signatures = multi_sig.signatures().to_vec();
+        signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
+        signatures.dedup();
+
+        // Verify signatures.
+        let mut signer_identities = self.verify_helper(message, &signatures)?;
+
+        // Sort and dedup the list of matched signer identities.
+        signer_identities.sort();
+        signer_identities.dedup();
+        Ok(signer_identities)
+    }
+
+    // The code that does the actual signature verification. We separate it from the
+    // public `verify` method so that we do not keep re-sorting an already
+    // sorted signers/signatures list.
+    fn verify_helper<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        signatures: &[SIG],
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
```

Comment:
Move these restrictions to `where` too
```suggestion
    fn verify_helper<SIG>(
        &self,
        message: &[u8],
        signatures: &[SIG],
    ) -> Result<Vec<S>, SignatureError>
    where
        SIG: Clone + Default + Digestible + Eq + Hash + Message + Ord + Serialize + Signature,
        S: Verifier<SIG>,
```